manifests/certificate/x509.pp

```diff
@@ -184,45 +170,45 @@
     extendedkeyusages => $extkeyusage,
     subjectaltnames   => $altnames,
   }
-  ~> x509_request { $_csr:
+  ~> x509_request { $csr:
     ensure      => $ensure,
-    template    => $_cnf,
-    private_key => $_key,
+    template    => $cnf,
+    private_key => $key,
     password    => $password,
     force       => $force,
     encrypted   => $encrypted,
   }
-  ~> x509_cert { $_crt:
+  ~> x509_cert { $crt:
     ensure   => $ensure,
-    template => $_cnf,
-    csr      => $_csr,
+    template => $cnf,
+    csr      => $csr,
     days     => $days,
     password => $password,
-    req_ext  => $req_ext,
+    req_ext  => !empty($altnames) and !empty($extkeyusage),
```

Comment:
@zilchms @rtib This condition seems to prevent ``altnames`` from being used. The condition used to be ``!empty($altnames + $extkeyusage)`` so either would trigger ``req_ext`` to be set and therefore the openssl command to be given the necessary option for the certificate to have the extension (this is done in ``lib/puppet/provider/x509_cert/openssl.rb`` line ``103``). This became a ``and`` condition which now requires both lists to be non-empty. I wasnt able to get certs with SANs to generate until I changed this locally to a ``or`` condition. I'll look into making a contribution shortly but I am commenting here just to have you take a look if you agree and maybe do the change if it's easy on your end.

Reply:
You're right, my mistake. AFAIK, I tested extkeyusage, which should work, but that can't work either. Unfortunately, the module lacks appropriate tests, which is a shame considering the importance of its functionality. I can file a PR, I broke it I fix it.

Reply:
Thanks for the quick answer! This is the basic change with a test (that fails with main branch but passes with change), but I'll let you decide if there is more is needed. I didnt see a test for the actual function generating the openssl command for instance but only did a quick glance.

```diff
diff --git a/manifests/certificate/x509.pp b/manifests/certificate/x509.pp
index 7543d7e..2aae7c4 100644
--- a/manifests/certificate/x509.pp
+++ b/manifests/certificate/x509.pp
@@ -184,7 +184,7 @@ define openssl::certificate::x509 (
     csr      => $csr,
     days     => $days,
     password => $password,
-    req_ext  => !empty($altnames) and !empty($extkeyusage),
+    req_ext  => !empty($altnames) or !empty($extkeyusage),
     force    => $force,
     ca       => $ca,
     cakey    => $cakey,
diff --git a/spec/defines/openssl_certificate_x509_spec.rb b/spec/defines/openssl_certificate_x509_spec.rb
index e2df33d..57facfe 100644
--- a/spec/defines/openssl_certificate_x509_spec.rb
+++ b/spec/defines/openssl_certificate_x509_spec.rb
@@ -422,6 +422,26 @@ describe 'openssl::certificate::x509' do
     }
   end

+  context 'when passing altnames, extension is enabled' do
+    let(:params) do
+      {
+        country: 'com',
+        organization: 'bar',
+        commonname: 'foo.example.com',
+        altnames: ['bar.example.com'],
+      }
+    end
+
+    it {
+      is_expected.to contain_x509_cert('/etc/ssl/certs/foo.crt').with(
+        ensure: 'present',
+        template: '/etc/ssl/certs/foo.cnf',
+        csr: '/etc/ssl/certs/foo.csr',
+        req_ext: true
+      )
+    }
+  end
+
   context 'when passing all parameters' do
     let(:params) do
       {
```

Reply:
Thanks @rtib , i will look at getting some tests up and running in the near future. This module really needs those....
and thanks @fmichea for reporting this :)

Reply:
Thanks @fmichea , took that, extended it with some more cases. @zilchms ,  #195 is ready for review.